### ISSUE TITLE
model.train() returns None when training on multiple GPUs.

### ISSUE THREAD

**Issue by xiaoyu-wen-1118** [ultralytics/ultralytics#19519]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training the YOLO model (both v11 and v12) on multiple GPUs (n_gpu > 1), the results returned from `model.train()` is always `None`.

Here’s an example code block I used to test this issue:

`results = model.train(
    data = "./dataset.yaml",
    epochs=10,
    fraction=0.2,
    batch=64,
    imgsz=128,
    conf=0.3,
    iou=0.2,
    device = [0,1],
)`

Here is a screenshot showing the type of `results` output:

![Image](https://github.com/user-attachments/assets/ff8d1e22-c328-4acd-896c-c8844dd658ce)

### Environment

Ultralytics 8.3.82 🚀 Python-3.11.10 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 14931MiB)
Setup complete ✅ (4 CPUs, 14.6 GB RAM, 34.2/94.3 GB disk)

OS                  Linux-6.1.123+-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.10
Install             pip
Path                /opt/conda/lib/python3.11/site-packages/ultralytics
RAM                 14.64 GB
Disk                34.2/94.3 GB
CPU                 Intel Xeon 2.30GHz
CPU count           4
GPU                 Tesla T4, 14931MiB
GPU count           2
CUDA                12.4

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1+cu124>=1.8.0
torch               ✅ 2.5.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1+cu124>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.1.4>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
{'OS': 'Linux-6.1.123+-x86_64-with-glibc2.35',
 'Environment': 'Linux',
 'Python': '3.11.10',
 'Install': 'pip',
 'Path': '/opt/conda/lib/python3.11/site-packages/ultralytics',
 'RAM': '14.64 GB',
 'Disk': '34.2/94.3 GB',
 'CPU': 'Intel Xeon 2.30GHz',
 'CPU count': 4,
 'GPU': 'Tesla T4, 14931MiB',
 'GPU count': 2,
 'CUDA': '12.4',
 'Package Info': {'numpy': '✅ 1.26.4<=2.1.1,>=1.23.0',
  'matplotlib': '✅ 3.9.2>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 10.4.0>=7.1.2',
  'pyyaml': '✅ 6.0.2>=5.3.1',
  'requests': '✅ 2.32.3>=2.23.0',
  'scipy': '✅ 1.14.1>=1.4.1',
  'torch': '✅ 2.5.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.20.1+cu124>=0.9.0',
  'tqdm': '✅ 4.67.1>=4.64.0',
  'psutil': '✅ 6.1.0',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.1.4>=1.1.4',
  'seaborn': '✅ 0.13.2>=0.11.0',
  'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}

### Minimal Reproducible Example

`results = model.train(
    data = "./dataset.yaml",
    epochs=10,
    fraction=0.2,
    batch=64,
    imgsz=128,
    conf=0.3,
    iou=0.2,
    device = [0,1],
)`
`print(type(results))`
`print(results.save_dir)`

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xiaoyu-wen-1118, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, thank you for providing detailed information about your issue. To assist in debugging further, could you also ensure that the provided code snippet aligns with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)? This helps us pinpoint the issue effectively.

If this is a custom training ❓ Question, please provide additional details, such as your dataset structure and specific logs during training, and check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

### Quick Suggestions

1. Ensure that your `ultralytics` package is up-to-date. Upgrade to the latest version with:
```bash
pip install -U ultralytics
```

2. Confirm that your environment (Python version, PyTorch, dependencies, etc.) meets the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). A [Python>=3.8](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) is recommended.

3. Ensure proper configuration for multi-GPU training. If there are any specific errors during training, sharing them would be helpful!

You can also try YOLO in different verified 💻 environments and platforms:

- **Notebooks** with free GPU access:  
<a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)

- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)

- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)  
<a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

You can also join the Ultralytics community for discussions and support:

- Discord 🎧: [Join here](https://discord.com/invite/ultralytics)
- Discourse Forum 📝: [Join here](https://community.ultralytics.com/)
- Subreddit 🦘: [Join here](https://reddit.com/r/Ultralytics)

### Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, confirming the model's functionality in verified environments.

Thank you again for bringing this to our attention! This is an automated response 🤖, but rest assured, an Ultralytics engineer will assist you as soon as possible. 😊

**Comment by glenn-jocher** [comment]
Thanks for reporting this. When using multiple GPUs, Ultralytics YOLO initiates distributed training via subprocesses, which means the main process doesn't return the training results object. This is expected behavior in multi-GPU configurations. Your training outputs (metrics, weights, plots) are still saved to `runs/train/` as usual.

For tracking training results programmatically in multi-GPU setups, you can access the saved metrics from the last saved results directory using:
```python
from ultralytics.utils import SETTINGS
print(SETTINGS['runs_dir'])
```

Note: The `fraction=0.2` parameter in your example isn't a standard YOLO train argument - you might want to verify if this is intended.

**Comment by Y-T-G** [comment]
It's because with multi-GPU, the training occurs in a different process. So the result object is not available in the original process.

**Comment by xiaoyu-wen-1118** [comment]
> Thanks for reporting this. When using multiple GPUs, Ultralytics YOLO initiates distributed training via subprocesses, which means the main process doesn't return the training results object. This is expected behavior in multi-GPU configurations. Your training outputs (metrics, weights, plots) are still saved to `runs/train/` as usual.
> 
> For tracking training results programmatically in multi-GPU setups, you can access the saved metrics from the last saved results directory using:
> 
> from ultralytics.utils import SETTINGS
> print(SETTINGS['runs_dir'])
> Note: The `fraction=0.2` parameter in your example isn't a standard YOLO train argument - you might want to verify if this is intended.

@glenn-jocher Thanks for your explanation! It would be helpful to include this instruction and code snippet in the YOLO documentation for other users.

**Comment by glenn-jocher** [comment]
Thank you for the documentation suggestion! We've added a note about multi-GPU training behavior to the [Train mode documentation](https://docs.ultralytics.com/modes/train/#multi-gpu-training) to help future users. The community-driven improvements like this help make YOLO better for everyone! 🚀

**Comment by curtis18** [comment]
@glenn-jocher Is it possible to save the best result to an excel file if using multiple GPU? Thank you.

The following code runs properly for single GPU, but the results become None when using multiple GPU.

```
    class_mAPs = results.box.maps
    class_names = results.names
    mAP_data = {
        'class_name': [class_names[int(cls)] for cls in range(len(class_mAPs))],
        'mAP': class_mAPs
    }
    mAP_df = pd.DataFrame(mAP_data)
    xlsx_path = os.path.join(args.output_dir, args.name, 'class_mAP_results.xlsx')
    mAP_df.to_excel(xlsx_path, index=False)
```

> Thanks for reporting this. When using multiple GPUs, Ultralytics YOLO initiates distributed training via subprocesses, which means the main process doesn't return the training results object. This is expected behavior in multi-GPU configurations. Your training outputs (metrics, weights, plots) are still saved to `runs/train/` as usual.
> 
> For tracking training results programmatically in multi-GPU setups, you can access the saved metrics from the last saved results directory using:
> 
> from ultralytics.utils import SETTINGS
> print(SETTINGS['runs_dir'])
> Note: The `fraction=0.2` parameter in your example isn't a standard YOLO train argument - you might want to verify if this is intended.

**Comment by glenn-jocher** [comment]
Yes, you can save class mAP results to Excel even when using multiple GPUs, but you'll need to access the saved metrics from the training directory rather than using the returned results object. Since multi-GPU training runs in subprocesses, you'll need to load the metrics after training completes.

Here's how you can adapt your code:

```python
import pandas as pd
from pathlib import Path
from ultralytics.utils import SETTINGS
import yaml

# After training completes
# Get the latest training run directory
runs_dir = Path(SETTINGS['runs_dir']) / 'train'
latest_run = sorted(runs_dir.glob('*'), key=lambda x: x.stat().st_mtime)[-1]

# Load metrics from the results.yaml file
metrics_path = latest_run / 'results.yaml'
with open(metrics_path) as f:
    metrics = yaml.safe_load(f)

# Extract class mAPs and names
class_mAPs = metrics['metrics/mAP50-95(B)_per_class']
class_names = metrics['names']

# Create DataFrame and save to Excel
mAP_data = {
    'class_name': [class_names[str(cls)] for cls in range(len(class_mAPs))],
    'mAP': class_mAPs
}
mAP_df = pd.DataFrame(mAP_data)
xlsx_path = latest_run / 'class_mAP_results.xlsx'
mAP_df.to_excel(xlsx_path, index=False)
```

This approach retrieves the metrics directly from the saved results file rather than relying on the returned object.

**Comment by curtis18** [comment]
@glenn-jocher Thanks for your help. The runs_dir = Path(SETTINGS['runs_dir']) / 'train' is 'runs/train', but there is no results.yaml file in the train directory. I also tried to change the project and name settings, and the result is the same.

<img width="686" alt="Image" src="https://github.com/user-attachments/assets/5ff09de2-8a2d-4ebc-bfaf-58a03fd271e2" />


<img width="164" alt="Image" src="https://github.com/user-attachments/assets/43be9041-b6e5-40ed-b876-6258ef160e51" />

**Comment by glenn-jocher** [comment]
@curtis18 You're right, I apologize for the confusion. When using multi-GPU training, the metrics are saved in a CSV file (results.csv) rather than YAML format. Here's the corrected approach to extract metrics after multi-GPU training:

```python
import pandas as pd
from pathlib import Path
from ultralytics.utils import SETTINGS

# After training completes
# Get the latest training run directory
runs_dir = Path(SETTINGS['runs_dir']) / 'train'
latest_run = sorted(runs_dir.glob('*'), key=lambda x: x.stat().st_mtime)[-1]

# Load metrics from the results.csv file
csv_path = latest_run / 'results.csv'
metrics_df = pd.read_csv(csv_path)

# Get the validation metrics from the final epoch
final_metrics = metrics_df.iloc[-1]

# For per-class metrics, you'll need to load the model and run validation again
model_path = latest_run / 'weights' / 'best.pt'
from ultralytics import YOLO
model = YOLO(model_path)
val_results = model.val()  # This will run on a single GPU

# Now you can create your Excel file
class_mAPs = val_results.box.maps
class_names = val_results.names
mAP_data = {
    'class_name': [class_names[cls] for cls in range(len(class_mAPs))],
    'mAP': class_mAPs
}
mAP_df = pd.DataFrame(mAP_data)
xlsx_path = latest_run / 'class_mAP_results.xlsx'
mAP_df.to_excel(xlsx_path, index=False)
```

This approach loads the trained model from the saved weights and runs validation once more (on a single GPU) to get the per-class metrics that you can then save to Excel.

**Comment by curtis18** [comment]
@glenn-jocher Thank you for your clarification and great help.

**Comment by glenn-jocher** [comment]
You're welcome @curtis18! Glad I could help. If you encounter any other issues with multi-GPU training or need further assistance with accessing metrics, feel free to reach out. The community's feedback helps us continuously improve the documentation and user experience for YOLO models.
